llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp

```diff
@@ -266,38 +266,67 @@ static Type *getPtrOrVecOfPtrsWithNewAS(Type *Ty, unsigned NewAddrSpace) {
+  if (!Op)
+    return nullptr;
+  bool validAddrSpaceCast = false;
+  Operator *P2I = nullptr;
   unsigned I2PAS = I2P->getType()->getPointerAddressSpace();
-  return CastInst::isNoopCast(Instruction::CastOps(I2P->getOpcode()),
-                              I2P->getOperand(0)->getType(), I2P->getType(),
-                              DL) &&
-         CastInst::isNoopCast(Instruction::CastOps(P2I->getOpcode()),
-                              P2I->getOperand(0)->getType(), P2I->getType(),
-                              DL) &&
-         (P2IOp0AS == I2PAS || TTI->isNoopAddrSpaceCast(P2IOp0AS, I2PAS));
+  switch (Op->getOpcode()) {
+  case Instruction::PtrToInt: {
+    // Check it's really safe to treat that pair of `ptrtoint`/`inttoptr` as a
+    // no-op cast. We need to double-check that through the target-specific
+    // hook. That ensures the underlying target also agrees that's a no-op
+    // address space cast and pointer bits are preserved. The current IR spec
+    // doesn't have clear rules on address space casts, especially a clear
+    // definition for pointer bits in non-default address spaces. It would be
+    // undefined if that pointer is dereferenced after an invalid reinterpret
+    // cast. Also, due to the unclearness for the meaning of bits in non-default
+    // address spaces in the current spec, the pointer arithmetic may also be
+    // undefined after invalid pointer reinterpret cast. However, as we confirm
+    // through the target hooks that it's a no-op addrspacecast, it doesn't
+    // matter since the bits should be the same.
+    P2I = Op;
+    unsigned P2IOp0AS = P2I->getOperand(0)->getType()->getPointerAddressSpace();
+    validAddrSpaceCast =
+        (P2IOp0AS == I2PAS || TTI->isNoopAddrSpaceCast(P2IOp0AS, I2PAS));
+    break;
+  }
+  case Instruction::ZExt: {
+    // check src address pointer and dst address pointer should be different
+    // address space and different size
+    P2I = dyn_cast<Operator>(Op->getOperand(0));
+    if (P2I && P2I->getOpcode() == Instruction::PtrToInt) {
+      unsigned P2IOp0AS =
+          P2I->getOperand(0)->getType()->getPointerAddressSpace();
+      unsigned P2IOp0SizeInBits =
+          DL.getPointerTypeSizeInBits(P2I->getOperand(0)->getType());
+      unsigned I2PSizeInBits = DL.getPointerTypeSizeInBits(I2P->getType());
+      validAddrSpaceCast =
+          (P2IOp0AS != I2PAS && P2IOp0SizeInBits < I2PSizeInBits);
+    }
+    break;
+  }
+  default:
+    break;
+  }
+  // We also need to check both `inttoptr` and `ptrtoint` are no-op casts, as
+  // the reinterpreted pointer may be used in other pointer arithmetic.
+  if (validAddrSpaceCast &&
+      CastInst::isNoopCast(Instruction::CastOps(I2P->getOpcode()),
+                           I2P->getOperand(0)->getType(), I2P->getType(), DL) &&
+      CastInst::isNoopCast(Instruction::CastOps(P2I->getOpcode()),
+                           P2I->getOperand(0)->getType(), P2I->getType(), DL)) {
+    return P2I;
```

Comment:
Make this a helper function called in the contexts above instead of falling through to post-process it?